vsphere/resource_vsphere_datacenter_test.go

```diff
@@ -0,0 +1,115 @@
+package vsphere
+
+import (
+	"fmt"
+	"testing"
+	"time"
+
+	"github.com/hashicorp/terraform/helper/resource"
+	"github.com/hashicorp/terraform/terraform"
+	"github.com/vmware/govmomi"
+	"github.com/vmware/govmomi/find"
+	"golang.org/x/net/context"
+)
+
+// Create a datacenter on the root folder
+func TestAccVSphereDatacenter_createOnRootFolder(t *testing.T) {
+	name := "testDC"
+	resourceName := "vsphere_datacenter." + name
+
+	waitForCreation := func(*terraform.State) error {
+		// sleep 5 seconds to give it time to create
+		time.Sleep(5 * time.Second)
+		return nil
+	}
+
+	resource.Test(t, resource.TestCase{
+		PreCheck:     func() { testAccPreCheck(t) },
+		Providers:    testAccProviders,
+		CheckDestroy: testAccCheckVSphereDatacenterDestroy,
+		Steps: []resource.TestStep{
+			{
+				Config: fmt.Sprintf(
+					testAccCheckVSphereDatacenterConfig,
+					name,
+					name,
+				),
+				Check: resource.ComposeTestCheckFunc(
+					waitForCreation,
```

Comment:
You are absolutely right, I missed this race condition. I have used ```WaitForState``` both for creation and deletion, I think it should be ok now.